### ISSUE TITLE
multi gpu training is worse than single gpu

### ISSUE THREAD

**Issue by xinqipony** [horovod/horovod#1198]
**Environment:**
1. Framework: (PyTorch)
2. Framework version:  
3. Horovod version: latest
4. MPI version:
5. CUDA version: 9
6. NCCL version:
7. Python version: 3
8. OS and version: linux
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
I am using horovod with pytorch, when I use 1 gpu , the model can converge well , but when I use 2 or 4, 8 gpu, the loss just becoming very large , compared to 1 gpu, I didn't change lr(lr * num_gpu doesn't work too), batch size,  just have more gpus, here's my code :

    train_dataloader = create_dataloader(data_config)
    model = create_model(model_config, train_dataloader)

    hvd.broadcast_parameters(model.state_dict(), root_rank=0)

	# adam
    optimizer = create_optimizer(optimizer_config, model)

    optimizer = hvd.DistributedOptimizer(optimizer, named_parameters=model.named_parameters())

    hvd.broadcast_optimizer_state(optimizer, root_rank=0)

	trainer = Trainer(model, optimizer, train_dataloader)
    trainer.run()
    

##########################
anyone knows what's going on here?

**Comment by alsrgv** [comment]
@xinqipony, this problem is hard to diagnose without knowing the model better. Do you see loss blowing up right from the beginning (first few iterations), or somewhere during the training? Did you try decreasing LR until loss does not explode?

**Comment by xinqipony** [comment]
thank you!

the model is just a seq2seq model with attention, problem is machine translation;

loss blows up after just 20 or 30 steps;

after decreasing lr, loss does not explode, but the loss decreases  too slow and can't convergence like one card

**Comment by alsrgv** [comment]
@xinqipony, could you collect [Horovod Timeline](https://horovod.readthedocs.io/en/latest/timeline.html)? It may help diagnose the problem futher.

**Comment by adriancaruana** [comment]
I am also seeing a similar problem. Whilst the loss isn't blowing up, the loss never achieves the same performance as single-gpu non-horovod experiments. I have tried experimenting with decreasing learning rate, but I see almost identical results regardless of LR (with ADAM). I'm using keras.

This experiment shows the binary-loss when training with 4x GPUs (Blue), 2x GPUs(Brown), and single-gpu (Pink) (no hvd). 

![Screen Shot 2019-07-10 at 4 22 23 pm](https://user-images.githubusercontent.com/38579731/60945498-1275a100-a32f-11e9-9dbf-44790f348df2.png)

Here's an excerpt of my horovod timeline:

![Screen Shot 2019-07-10 at 4 48 58 pm](https://user-images.githubusercontent.com/38579731/60947114-215e5280-a333-11e9-97f2-0caef829c209.png)

**Comment by xinqipony** [comment]
another thing is that for one gpu training and multi gpu training I used horovod both, the only difference is the gpu number;

I tried another distributed training using pytorch chainer, the issue seems gone;

so I'm wondering if I misuses horovod for multi gpu;

**Comment by adriancaruana** [comment]
Update: I've tried training with single-gpu but with 1/4 of my train-set (akin to what a single GPU would see when training 4x multi-gpu with sharding). Note I've seen this issue using CUDA9 and 10, and TF 1.12 and 1.14. 

The new blue line (the longest one) is the single-gpu with 1/4 of train-set:

![Screen Shot 2019-07-11 at 10 08 29 am](https://user-images.githubusercontent.com/38579731/61012983-2ae3cb00-a3c4-11e9-8a18-120ea9dccb6e.png)

This leads me to believe that there is perhaps something going wrong with the all-reduce operation. 

Also note, that (I believe) I am creating the callbacks/optimisers appropriately for horovod multi-gpu training. Here is the general structure of my code:
```
opt = Adam(lr=ADAM_LR*hvd.size())
opt = hvd.DistributedOptimizer(opt)
model.compile(..., optimizer=opt, ...)
...
callbacks = []
callbacks.append(hvd.callbacks.BroadcastGlobalVariablesCallback(0))
callbacks.append(hvd.callbacks.MetricAverageCallback())
if hvd.rank() == 0:
    callbacks.append(tensorboard)
...
model.fit_generator(..., callbacks=callbacks, ...)
```

**Comment by Oktai15** [comment]
@alsrgv, I have the same problem with TF 2.0 -- loss is becoming super large when I use multi-gpu.

**Comment by xinqipony** [comment]
I am too confused, as my understanding, multi gpu training should be faster and more stable when using horovod; 

I am testing another framework called chainer in pytorch for multi gpu training, it seems chainer doesn't have this issue;

**Comment by Oktai15** [comment]
My tracing looks like that (I can't understand how to zoom it with labels of bars)

![image](https://user-images.githubusercontent.com/17337773/61028448-6204ad00-a3c1-11e9-9f22-23ab2c7206cb.png)

**Comment by Oktai15** [comment]
I fixed my large loss in multi-gpu mode. I left the same learning rate that I have in one gpu mode (it means that I didn't scale lr via `hvd.size()`). I don't know why but it returns me to the same loss values.

Maybe it is problem with new approach to use horovod in TF 2.0 when we use `tf.keras.optimizers.Adam`, i.e no wrapper for optimizer as earlier, and wrapper for gradient tape, i.e `hvd.DistributedGradientTape(tape)`. Should we wrap optimizer in something like before, @alsrgv? According to your new examples, we don't need, but..?

**Comment by xinqipony** [comment]
though lower lr makes the loss value seems the same, but in my experience, the model accuracy can't be comparable to one card training model

**Comment by chenghuige** [comment]
I face the same problem. With one gpu batch size 512, I want to check if using horovod 8 gpu each batch size 64 will reproduce same result (all other things the same, like lr..).
For each batch 64 worker shows lower loss then 1 gpu training. (0.3 to 0.29)
But I got worse valid loss and auc .

**Comment by kaituoxu** [comment]
Any update? Watching...

**Comment by tgaddair** [comment]
Hey @mariamrahmaani, there are many things that can cause poor scaling ranging from environment setup bugs, or slow networks to problems with the training script.  Feel free to open an new issue with your specific details and we'd be happy to help you diagnose any problems.

**Comment by jusonn** [comment]
does this issue solved??

**Comment by tgaddair** [comment]
Hey @jusonn, which issue are you referring to?  If you're seeing that multi gpu training is worse than single gpu, then most likely it's an issue with the specific model or environment.  Feel free to create a new issue so we can dive into the issue further.

**Comment by marcelgwerder** [comment]
@adriancaruana Ever figured out what your issue was exactly? I see similar behavior with my PyTorch code. My training also looks like I'm only training on 1/4 of the dataset and I am also using Adam.

**Comment by zfxxfeng** [comment]
I use hovord +estimator( tf.estimator.train_and_evaluate(classifier, train_spec, eval_spec)). I get a strange phenomenon. My train_acc which log in tensorboard is high but the eval_acc is low. And I test the train data in another script ,the acc is also low.I use lr* hvd.size()(32gpu)
![image](https://user-images.githubusercontent.com/22147218/76180147-1d533f00-61f8-11ea-9ce3-2044013654b6.png)

![image](https://user-images.githubusercontent.com/22147218/76180062-da916700-61f7-11ea-832c-61cc072b4adf.png)
When I use (optimizer = tf.train.AdamOptimizer(learning_rate=0.001 )),Looks normal.I am very confused.

**Comment by tgaddair** [comment]
Hey @zfxxfeng, using `estimator.train_and_evaluate` requires special consideration.  Are you setting `throttle_secs=0` in your `EvalSpec`?

**Comment by zfxxfeng** [comment]
Hello@tgaddair  I already set throttle_secs=0.But the loss and eval_acc still shock.

**Comment by talkhaldi** [comment]
Has this been solved?

I'm having a similar issue. My learning rate is 5e-5. when I multiply with hvd.size() (16 in my case), the loss doesn't decrease. If I add warmup steps 0.1 of total steps, the loss decreases while the lr is still low, but when the warmup phase is over halfway, the loss starts increasing again. It seems if I don't multiply with hvd.size() it would work well, but idk if this is the right way...

**Comment by talkhaldi** [comment]
Okay, I found out what was wrong. The learning rate I was using was 5e-5 for a 32 batch. I've used  it with DistributedDataParallel on single machine, multi-GPU setup. In Horovod, each worker on a separate GPU has its own "local" batch size, around 4 per GPU. 

Even though DistributedDataParallel also spawns separate workers with local "per GPU", the learning rate was applied to the whole 32 batch (or was split internally, I'm not really sure).

With Horovod, I had to divide the learning rate as follows:

```
args.learning_rate = args.learning_rate / hvd.size() * (args.effective_batch_size / 32) #Use learning rates that would work for 32 batch and scale them

lr_scaler = hvd.size() if not args.use_adasum else 1
if args.use_adasum and hvd.nccl_built():
     lr_scaler = hvd.local_size()

optimizer = AdamW(optimizer_grouped_parameters, lr=args.learning_rate * lr_scaler, eps=args.adam_epsilon)
```

Hope it helps someone!

**Comment by iodapro** [comment]
@talkhaldi ,
I didn't get what is your args.effective_batch_size?

**Comment by talkhaldi** [comment]
It's the actual batch size that will affect the convergence. This effective_batch_size will be split to local_batch_size per worker, depending on hvd.size().

**Comment by iodapro** [comment]
I have similar problem, that 3 GPU loss is higher than the one used in single GPU regime.
So, suppose I have 3 GPUs, then hvd.size() is 3 and effective_batch_size can be increased in 3 times.
If in single GPU regime I used lr = 3e-4 and batch_size is 16, then with 3 GPUs it is 
3e-4/hvd.size() and effective_batch_size is 16*3
or
3e-4 / hvd.size() * 16*3 / 16 = 3e-4 and effective_batch_size is 16*3

I am so confused with it. I used guidelines like in horovod example, but have the same problem.

**Comment by talkhaldi** [comment]
So basically the effective batch size you want is 16 * 3 = 48, right?

The learning rate should be multiplied by 3 then. If you input it through the code above it will become:

3e-4 / hvd.size() * (16 * 3 / 16) = 3e-4 (per worker, in your case it's the same because the workers still have the same batch size (16))

Then lr_scaler = hvd.size()

and your final lr = 3e-4 * lr_scaler, which is just the normal lr * 3 as it should be.

In my case the formulas were necessary because the workers ended up having lower local batch.
